osquery/logger/plugins/kinesis.cpp

```diff
@@ -0,0 +1,125 @@
+#include <aws/kinesis/model/PutRecordsRequest.h>
+#include <aws/kinesis/model/PutRecordsRequestEntry.h>
+#include <aws/kinesis/model/PutRecordsResult.h>
+
+#include <osquery/flags.h>
+#include <osquery/registry.h>
+
+#include "osquery/logger/plugins/aws_util.h"
+#include "osquery/logger/plugins/kinesis.h"
+
+namespace osquery {
+
+REGISTER(KinesisLoggerPlugin, "logger", "kinesis");
+
+FLAG(uint64,
+     logger_kinesis_period,
+     10,
+     "Seconds between flushing logs to Kinesis (default 10)");
+FLAG(string, aws_kinesis_stream, "", "Name of Kinesis stream for logging")
+
+const size_t KinesisLogForwarder::kKinesisMaxRecords = 500;
+// Max size of log + partition key is 1MB. Max size of partition key is 256B.
+const size_t KinesisLogForwarder::kKinesisMaxLogBytes = 1000000 - 256;
+
+Status KinesisLoggerPlugin::setUp() {
+  forwarder_ = std::make_shared<KinesisLogForwarder>();
+  Status s = forwarder_->setUp();
+  if (!s.ok()) {
+    return s;
+  }
+  Dispatcher::addService(forwarder_);
+  return Status(0, "OK");
+}
+
+Status KinesisLoggerPlugin::logString(const std::string& s) {
+  return forwarder_->logString(s);
+}
+
+Status KinesisLogForwarder::send(std::vector<std::string>& log_data,
+                                 const std::string& log_type) {
+  std::vector<Aws::Kinesis::Model::PutRecordsRequestEntry> entries;
+  for (const std::string& log : log_data) {
+    if (log.size() > kKinesisMaxLogBytes) {
+      LOG(ERROR) << "Kinesis log too big, discarding!";
+    }
+    Aws::Kinesis::Model::PutRecordsRequestEntry entry;
+    entry.WithPartitionKey(shard_id_).WithData(
+        Aws::Utils::ByteBuffer((unsigned char*)log.c_str(), log.length()));
+    entries.push_back(std::move(entry));
+  }
+
+  Aws::Kinesis::Model::PutRecordsRequest request;
+  request.WithStreamName(FLAGS_aws_kinesis_stream)
+      .WithRecords(std::move(entries));
+
+  Aws::Kinesis::Model::PutRecordsOutcome outcome = client_->PutRecords(request);
+  Aws::Kinesis::Model::PutRecordsResult result = outcome.GetResult();
+  if (result.GetFailedRecordCount() != 0) {
+    for (const auto& record : result.GetRecords()) {
+      if (!record.GetErrorMessage().empty()) {
+        LOG(ERROR) << "Kinesis write for " << result.GetFailedRecordCount()
+                   << " of " << result.GetRecords().size()
+                   << " records failed with error " << record.GetErrorMessage();
+        return Status(1, record.GetErrorMessage());
+      }
+    }
+  }
+
+  VLOG(1) << "Successfully sent " << result.GetRecords().size()
+          << " logs to Kinesis.";
+  return Status(0, "OK");
+}
+
+Status KinesisLogForwarder::setUp() {
+  // Set up client
+  Status s = makeAWSClient<Aws::Kinesis::KinesisClient>(client_);
+  if (!s.ok()) {
+    return s;
+  }
+
+  shard_id_ = getHostIdentifier();
+
+  if (FLAGS_aws_kinesis_stream.empty()) {
+    std::string err = "Stream name must be specified with --aws_kinesis_stream";
+    LOG(ERROR) << err;
+    return Status(1, err);
```

Comment:
For these (1) construct string, (2) log string, (3) set string as status message steps, maybe just print the error here: https://github.com/facebook/osquery/pull/2045/files#diff-a742a9993ebc7ea2aca7f460f21cfea4R46 instead.